flow/layers/picture_raster_cache_item.cc

```diff
@@ -0,0 +1,172 @@
+
+  if (picture == nullptr ||
+      !RasterCacheUtil::CanRasterizeRect(picture->cullRect())) {
+    // No point in deciding whether the picture is worth rasterizing if it
+    // cannot be rasterized at all.
+    return false;
+  }
+
+  if (is_complex) {
+    // The caller seems to have extra information about the picture and thinks
+    // the picture is always worth rasterizing.
+    return true;
+  }
+
+  // TODO(abarth): We should find a better heuristic here that lets us avoid
+  // wasting memory on trivial layers that are easy to re-rasterize every frame.
+  return picture->approximateOpCount(true) > 5;
+}
+
+SkPictureRasterCacheItem::SkPictureRasterCacheItem(SkPicture* sk_picture,
+                                                   const SkPoint& offset,
+                                                   bool is_complex,
+                                                   bool will_change)
+    : RasterCacheItem(RasterCacheKeyID(sk_picture->uniqueID(),
+                                       RasterCacheKeyType::kPicture),
+                      CacheState::kCurrent),
+      sk_picture_(sk_picture),
+      offset_(offset),
+      is_complex_(is_complex),
+      will_change_(will_change) {}
+
+std::unique_ptr<SkPictureRasterCacheItem> SkPictureRasterCacheItem::Make(
+    SkPicture* sk_picture,
+    const SkPoint& offset,
+    bool is_complex,
+    bool will_change) {
+  return std::make_unique<SkPictureRasterCacheItem>(sk_picture, offset,
+                                                    is_complex, will_change);
+}
+
+void SkPictureRasterCacheItem::PrerollSetup(PrerollContext* context,
+                                            const SkMatrix& matrix) {
+  cache_state_ = CacheState::kNone;
+  if (!IsPictureWorthRasterizing(sk_picture_, will_change_, is_complex_)) {
+    // We only deal with pictures that are worthy of rasterization.
+    return;
+  }
+  transformation_matrix_ = matrix;
+  transformation_matrix_.preTranslate(offset_.x(), offset_.y());
+  if (!transformation_matrix_.invert(nullptr)) {
+    // The matrix was singular. No point in going further.
+    return;
+  }
+  if (context->raster_cached_entries && context->raster_cache) {
+    context->raster_cached_entries->push_back(this);
+    cache_state_ = CacheState::kCurrent;
+  }
+
+  return;
+}
+
+void SkPictureRasterCacheItem::PrerollFinalize(PrerollContext* context,
+                                               const SkMatrix& matrix) {
+  if (cache_state_ == CacheState::kNone || !context->raster_cache ||
+      !context->raster_cached_entries) {
+    return;
+  }
+  auto* raster_cache = context->raster_cache;
+  SkRect bounds = sk_picture_->cullRect().makeOffset(offset_.x(), offset_.y());
+  // We've marked the cache entry that we would like to cache so it stays
+  // alive, but if the following conditions apply then we need to set our
+  // state back to !cacheable_ so that we don't populate the entry later.
+  if (!SkRect::Intersects(context->cull_rect, bounds)) {
```

Comment:
Same comments about the interaction of this line with the MarkSeen below as I described in the DisplayList version.